stm32: tests: drivers: spi: spi_loopback: add support for new spi test cases

djiatsaf-st opened the pull request:
This PR introduces updates on STM32 boards related to #87838, which adds new test cases (specifically: different word size tests, timing test cases, and CS testing) for the SPI loopback test driver.

That said, there is still a remaining issue where the testsuite `test_spi_hold_on_cs` fails on the nucleo_h743zi board only.

Test result logs are blocked. It looks like the board doesn't support the SPI operation configuration :
`hold_spec->config.operation |= SPI_HOLD_ON_CS`

djiatsaf-st committed:
tests: drivers: spi_loopback: add latency tolerance for some boards

Set a convenient latency for each platform, since each platform
reacts differently to the driver.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
tests: drivers: spi: spi_loopback:  enable nucleo_c071rb to run test again

Due to RAM overflow by 3200 bytes, decrease size of SPI_LARGE_BUFFER.

signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st commented on tests/drivers/spi/spi_loopback/boards/nucleo_f746zg.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_NOCACHE_MEMORY=y
+CONFIG_DCACHE=n
```
Following suggestion i had here https://github.com/zephyrproject-rtos/zephyr/pull/90321 

Initially, my intention was to update the SPI driver at https://github.com/zephyrproject-rtos/zephyr/blob/6559c7f5547733692d0158f2d1b7205f1c6db5bd/drivers/spi/spi_ll_stm32.c#L1145, changing EFAULT to ENOSUP to allow skipping the test_spi_word_size_x testcases.

However, I realized there was still an issue with the test_spi_concurrent_transfer_same_spec and test_spi_concurrent_transfer_different_spec test cases. The problem is that `tx_buffer` and `rx_buffer` do not have the `__NOCACHE` attribute. I assume this is intentional by the test writer, so I decided to disable D-cache for some ST boards to avoid causing regressions on other vendor platforms.

Changing `EFAULT` to `ENOSUP` and adding the `__NOCACHE` attribute to `tx_buffer` and `rx_buffer` could resolve the regression without disabling D-cache, but it might lead to regressions on other vendor platforms.

djiatsaf-st committed:
tests: drivers: spi: spi_loopback: handle rx/tx buffers nocache memory

Place transfer buffers in non-cacheable memory when
CONFIG_NOCACHE_MEMORY=y. This change ensures that DMA transfer
buffers are allocated in non-cacheable memory on platforms where
CONFIG_NOCACHE_MEMORY is enabled. This avoids potential cache
coherence issues that are not handled by the SPI driver.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
drivers: spi: update error type to skip word_size testcases

Workaround to avoid getting the -EFAULT (14) error when running
the test_spi_word_size_x (x = 7, 9, 24) test cases:
Since these testcase configurations aren't supported, update
the error type to ENOTSUP to allow skipping the test cases using ztest.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

JarmouniA reviewed (commented):
the 3rd commit's message needs reword

djiatsaf-st commented:
> the 3rd commit's message needs reword

Thanks

gautierg-st commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -621,7 +621,11 @@ static void spi_loopback_test_word_size(struct spi_dt_spec *spec,
 
 	config_copy.operation &= ~SPI_WORD_SIZE_MASK;
-	config_copy.operation |= SPI_WORD_SET(word_size);
+	if (word_size == 16) {
+		config_copy.operation |= SPI_WORD_SET(word_size - 1);
+	} else {
+		config_copy.operation |= SPI_WORD_SET(word_size);
+	}
```
This commit should be dropped. That may make the STM32 tests pass but it hides the problem.
STM32 supports 16-bit SPI transfer. And `test_spi_word_size_16` function passes when polling or interrupt transfers are used.

The problem lies when we activate the DMA. DMA is configured in dts overlays, and the size is defined there also. So when the DMA is configured for 8-bit transfers, the 16-bit test fails. For the 16-bit test to pass, DMA should be configured as 16-bit, but then the 8-bit test would fail. 

Until recently we relied on `CONFIG_SPI_LOOPBACK_16BITS_FRAMES` to know how to configure the DMA. But with the new tests, it isn't sufficient.

@decsny You introduced the new `test_spi_word_size_16` test. Do you have an idea how to solve this?

decsny replied:
this is not correct, there shouldn't be something special in the API about 16 bits, we want to test it, not 15 bits. If your implementation is not passing, it's not a fault of the test

djiatsaf-st committed:
drivers: spi: stm32: dynamically update spi word/frame size

Update the source_data and dest_data sizes with the frame size
in bytes before performing the transfer.
With these changes, we can handle cases where the buffers switch
between 8-bit and 16-bit modes.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
tests: drivers: spi: remove useless files and testcases

- with the updates made to the spi_ll_stm32 driver,
we no longer need to configure the 16-bit frame/word size
in the DTS overlay file. The test_spi_word_size_16 testsuite
helps us verify the 16-bit frame mode supported on
the ST platform.

- remove testcases that use thoses files in testcase.yaml

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st commented on tests/drivers/spi/spi_loopback/src/spi.c:
Done

etienne-lms commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -1186,6 +1186,14 @@ static int transceive_dma(const struct device *dev,
 	/* This is turned off in spi_stm32_complete(). */
 	spi_stm32_cs_control(dev, true);
 
+	uint8_t word_size_bytes = bits2bytes(
+			SPI_WORD_SIZE_GET(config->operation));
```
Nitpicking: fits in a single line:
```suggestion
	uint8_t word_size_bytes = bits2bytes(SPI_WORD_SIZE_GET(config->operation));

etienne-lms reviewed (commented):
In commit "tests: drivers: spi: spi_loopback: enable nucleo_c071rb to run test a gain",
remind to remove
```
- Add configuration for latency tolerance
```
from the commit message.
